### PR TITLE
chore: Use msgpack to reduce redundancy

### PR DESCRIPTION
**Pull request by ludamad**
# Description

Going through lists-of-fields now that we have poor-man's reflection.

This currently DOES NOT switch barretenberg's serialization to msgpack. This instead uses msgpack common reflection utils to derive such read and write functions.

For example:
MSGPACK_FIELDS(arg1, arg2);

is now used to define
```
void read(buf, obj) {
   read(buf, obj.arg1);
   read(buf, obj.arg2);
}
```
for any object that has a MSGPACK_FIELDS macro. This is also used for write

Also removes a debug utility I introduced during the hackathon that has not really been used, fine to remove

# Checklist:

- [*] I have reviewed my diff in github, line by line.
- [*] Every change is related to the PR description.
- [*] The branch has been merged with/rebased against the head of its merge target.
- [*] There are no unexpected formatting changes, superfluous debug logs, or commented-out code.
- [*] There are no circuit changes, OR a cryptographer has been assigned for review.
- [*] New functions, classes, etc. have been documented according to the doxygen comment format. Classes and structs must have `@brief` describing the intended functionality.
- [*] If existing code has been modified, such documentation has been added or updated.
- [*] No superfluous `include` directives have been added.
- [*] I have [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) to any issue(s) it resolves.
- [*] I'm happy for the PR to be merged at the reviewer's next convenience.

### CODE REVIEW

File: cpp/src/barretenberg/serialize/msgpack_impl/func_traits.hpp [hunk 1]
```diff
@@ -7,15 +7,15 @@ template <typename Func> struct func_traits;
 
 // Specialization for function pointers
 template <typename R, typename... Vs> struct func_traits<R (*)(Vs...)> {
-    typedef std::tuple<Vs...> Args; // Define a tuple type that holds all argument types
-    Args args;                      // Args instance
-    R ret;                          // Holds return type
-    MSGPACK_FIELDS(args, ret);      // Macro from msgpack library to serialize/deserialize fields
+    typedef std::tuple<typename std::decay<Vs>::type...> Args; // Define a tuple type that holds all argument types
```
Comment (ludamad) [inline cpp/src/barretenberg/serialize/msgpack_impl/func_traits.hpp:10]: Needed to support const Type& args

File: cpp/src/barretenberg/dsl/acir_format/ecdsa_secp256r1.test.cpp [hunk 2]
```diff
@@ -120,24 +120,25 @@ TEST(ECDSASecp256r1, test_hardcoded)
 
     size_t num_variables =
         generate_r1_constraints(ecdsa_r1_constraint, witness_values, pub_key_x, pub_key_y, hashed_message, signature);
+
     acir_format constraint_system{
         .varnum = static_cast<uint32_t>(num_variables),
         .public_inputs = {},
```
Comment (ludamad) [inline cpp/src/barretenberg/dsl/acir_format/ecdsa_secp256r1.test.cpp:126]: Reordered to match serialization order (hard to work around doing this)

File: cpp/src/barretenberg/common/serialize.hpp [hunk 3]
```diff
@@ -42,7 +44,14 @@
 __extension__ using uint128_t = unsigned __int128;
 #endif
 
+template <typename T>
+concept IntegralOrEnum = std::integral<T> || std::is_enum_v<T>;
```
Comment (spalladino) [inline cpp/src/barretenberg/common/serialize.hpp:48]: Just when I thought I was catching up with cpp...

File: cpp/src/barretenberg/common/serialize.hpp [hunk 4]
```diff
@@ -308,29 +288,45 @@ template <typename B> inline void read(B& it, std::string& value)
...
+    using serialize::write;
     write(buf, value.first);
     write(buf, value.second);
 }
 
+// Read std::shared_ptr.
+template <typename B, typename T> inline void read(B& it, std::shared_ptr<T>& value_ptr)
+{
+    using serialize::read;
+    T value;
```
Comment (spalladino) [inline cpp/src/barretenberg/common/serialize.hpp:315]: Why is this line needed?

File: cpp/src/barretenberg/crypto/schnorr/c_bind.cpp [hunk 5]
```diff
@@ -80,8 +80,8 @@ WASM_EXPORT void multisig_construct_signature_round_1(uint8_t* round_one_public_
     using multisig = crypto::schnorr::multisig<grumpkin::g1, KeccakHasher, Blake2sHasher>;
 
     auto [public_output, private_output] = multisig::construct_signature_round_1();
-    write(round_one_public_output_buf, public_output);
-    write(round_one_private_output_buf, private_output);
+    serialize::write_msgpack(round_one_public_output_buf, public_output);
+    serialize::write(round_one_private_output_buf, private_output);
```
Comment (spalladino) [inline cpp/src/barretenberg/crypto/schnorr/c_bind.cpp]: Why write_msgpack for one and write for the other?

**Review by spalladino — COMMENTED** [review]
LGTM, but don't count it as an approval since much of the cpp magic used here still escapes me.

File: cpp/src/barretenberg/crypto/schnorr/c_bind.cpp (hunk 5 above)
Comment (ludamad) [reply cpp/src/barretenberg/crypto/schnorr/c_bind.cpp]: I'll try undoing this hack now, thanks for the reminder

File: cpp/src/barretenberg/common/serialize.hpp (hunk 4 above)
Comment (ludamad) [reply cpp/src/barretenberg/common/serialize.hpp:315]: Good eye :) unfinished code

**Comment by ludamad** [comment]
> LGTM, but don't count it as an approval since much of the cpp magic used here still escapes me.

Foiled again! It's almost like I should avoid C++ magic

File: cpp/src/barretenberg/common/serialize.hpp [hunk 6]
```diff
@@ -449,4 +446,57 @@ using out_str_buf = uint8_t**;
...
+ * @param func The function to call with each field as an argument.
+ */
+inline void write(auto& buf, const msgpack_concepts::HasMsgPack auto& obj)
+{
+    using namespace serialize;
+    msgpack::msgpack_apply(obj, [&](auto&... obj_fields) {
+        // apply 'write' to each object field
+        (write_msgpack_field(buf, obj_fields), ...);
+    });
+};
```
Comment (dbanks12) [inline cpp/src/barretenberg/common/serialize.hpp]: cool

File: cpp/src/barretenberg/common/serialize.hpp [hunk 7]
```diff
@@ -449,4 +446,57 @@ using out_str_buf = uint8_t**;
...
+ * @param it The iterator to read from.
+ * @param func The function to call with each field as an argument.
+ */
+inline void read(auto& it, msgpack_concepts::HasMsgPack auto& obj)
+{
+    msgpack::msgpack_apply(obj, [&](auto&... obj_fields) {
+        // apply 'read' to each object field
+        read_msgpack_fields(it, obj_fields...);
+    });
+};
```
Comment (dbanks12) [inline cpp/src/barretenberg/common/serialize.hpp]: cool

**Comment by dbanks12** [comment]
Cannot parse "This currently DOES NOT switch barretenberg's serialization to msgpack. This instead uses msgpack common reflection utils to derive such read and write functions."

Could you explain this like im a 5 year old?

File: cpp/src/barretenberg/dsl/acir_format/acir_format.hpp [hunk 8]
```diff
@@ -23,24 +24,43 @@ struct acir_format {
...
     std::vector<EcdsaSecp256r1Constraint> ecdsa_r1_constraints;
-    std::vector<Sha256Constraint> sha256_constraints;
     std::vector<Blake2sConstraint> blake2s_constraints;
     std::vector<KeccakConstraint> keccak_constraints;
     std::vector<KeccakVarConstraint> keccak_var_constraints;
-    std::vector<HashToFieldConstraint> hash_to_field_constraints;
     std::vector<PedersenConstraint> pedersen_constraints;
-    std::vector<BlockConstraint> block_constraints;
+    std::vector<HashToFieldConstraint> hash_to_field_constraints;
+    std::vector<FixedBaseScalarMul> fixed_base_scalar_mul_constraints;
```
Comment (dbanks12) [inline cpp/src/barretenberg/dsl/acir_format/acir_format.hpp:38]: why the reordering?

File: cpp/src/barretenberg/dsl/acir_format/acir_format.test.cpp [hunk 9]
```diff
@@ -25,21 +25,21 @@ TEST(acir_format, test_a_single_constraint_no_pub_inputs)
...
         .keccak_constraints = {},
         .keccak_var_constraints = {},
-        .hash_to_field_constraints = {},
         .pedersen_constraints = {},
-        .block_constraints = {},
+        .hash_to_field_constraints = {},
+        .fixed_base_scalar_mul_constraints = {},
         .recursion_constraints = {},
         .constraints = { constraint },
+        .block_constraints = {},
```
Comment (dbanks12) [inline cpp/src/barretenberg/dsl/acir_format/acir_format.test.cpp:42]: same as above (why the reordering)?

File: cpp/src/barretenberg/dsl/acir_format/acir_format.test.cpp [hunk 10]
```diff
@@ -297,7 +291,9 @@ TEST(acir_format, test_schnorr_verify_small_range)
             .q_o = 1,
             .q_c = fr::neg_one(),
         } },
+        .block_constraints = {},
```
Comment (dbanks12) [inline cpp/src/barretenberg/dsl/acir_format/acir_format.test.cpp:294]: again, why reordering?

File: cpp/src/barretenberg/dsl/acir_format/acir_to_constraint_buf.hpp [hunk 11]
```diff
@@ -129,22 +129,23 @@ void handle_blackbox_func_call(Circuit::Opcode::BlackBoxFuncCall const& arg, aci
...
+                    .result = arg.output.value,
+                });
             } else if constexpr (std::is_same_v<T, Circuit::BlackBoxFuncCall::EcdsaSecp256k1>) {
                 af.ecdsa_k1_constraints.push_back(EcdsaSecp256k1Constraint{
                     .hashed_message = map(arg.hashed_message, [](auto& e) { return e.witness.value; }),
+                    .signature = map(arg.signature, [](auto& e) { return e.witness.value; }),
                     .pub_x_indices = map(arg.public_key_x, [](auto& e) { return e.witness.value; }),
                     .pub_y_indices = map(arg.public_key_y, [](auto& e) { return e.witness.value; }),
                     .result = arg.output.value,
-                    .signature = map(arg.signature, [](auto& e) { return e.witness.value; }),
```
Comment (dbanks12) [inline cpp/src/barretenberg/dsl/acir_format/acir_to_constraint_buf.hpp:147]: reorder?

File: cpp/src/barretenberg/dsl/acir_format/acir_to_constraint_buf.hpp [hunk 12]
```diff
@@ -180,8 +181,8 @@ void handle_blackbox_func_call(Circuit::Opcode::BlackBoxFuncCall const& arg, aci
                                           .num_bits = e.num_bits,
                                       };
                                   }),
-                    .var_message_size = arg.var_message_size.witness.value,
                     .result = map(arg.outputs, [](auto& e) { return e.value; }),
+                    .var_message_size = arg.var_message_size.witness.value,
```
Comment (dbanks12) [inline cpp/src/barretenberg/dsl/acir_format/acir_to_constraint_buf.hpp:185]: sorry, for being repetitive.... just want to make sure i mention the reordering wherever i see it

File: cpp/src/barretenberg/dsl/acir_format/acir_format.hpp (hunk 8 above)
Comment (dbanks12) [reply cpp/src/barretenberg/dsl/acir_format/acir_format.hpp:38]: saw later comment explaining. will resolve all my other comments reordering

File: cpp/src/barretenberg/serialize/msgpack_apply.hpp [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+#include "msgpack.hpp"
+#include "msgpack_impl/drop_keys.hpp"
+
+namespace msgpack {
+/**
+ * @brief Passthrough method for better error reporting.
+ */
+template <msgpack_concepts::HasMsgPack T> void msgpack_apply(const auto& func, auto&... args)
+{
+    std::apply(func, msgpack::drop_keys(std::tie(args...)));
```
Comment (dbanks12) [inline cpp/src/barretenberg/serialize/msgpack_apply.hpp]: Can you expand on this comment a bit (or the fn below)? What does it mean that they are passthrough methods and how might they be used?

File: cpp/src/barretenberg/serialize/msgpack_impl/drop_keys.hpp [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+/**
+ * @brief Drops every first value pairwise of a flat argument tuple, assuming that they are keys.
+ */
+template <typename... Args> auto drop_keys(std::tuple<Args...>&& tuple)
+{
+    static_assert(sizeof...(Args) % 2 == 0, "Tuple must contain an even number of elements");
+    // Compile time sequence of integers from 0 to n/2
+    auto compile_time_0_to_n_div_2 = std::make_index_sequence<sizeof...(Args) / 2>{};
+    return drop_keys_impl(tuple, compile_time_0_to_n_div_2);
+}
```
Comment (dbanks12) [inline cpp/src/barretenberg/serialize/msgpack_impl/drop_keys.hpp:21]: Why did you need to add these drop keys functions?

Comment (dbanks12) [reply cpp/src/barretenberg/serialize/msgpack_impl/drop_keys.hpp:21]: oh i see you just moved them

**Review by dbanks12 — APPROVED** [review]
LGTM, left some minor comments.

**Comment by ludamad** [comment]
> Cannot parse "This currently DOES NOT switch barretenberg's serialization to msgpack. This instead uses msgpack common reflection utils to derive such read and write functions."
> 
> Could you explain this like im a 5 year old?

I hope the edited text now makes more sense
